lib/elixir/lib/calendar/datetime.ex

```diff
@@ -1674,6 +1676,131 @@ defmodule DateTime do
     end
   end
 
+  @doc """
+  Shifts given `datetime` by `duration` according to its calendar.
+
+  Allowed units are: `:year, :month, :week, :day, :hour, :minute, :second, :microsecond`.
+
+  First the datetime is converted to a naive datetime. After the shift was applied
+  it is converted back to a datetime using its original time zone and time zone database,
+  potentially resulting in an ambiguous or gap DateTime result tuple.
+
+  Check `from_naive/3` for more information on ambiguous datetimes.
+
+  When used with the default calendar `Calendar.ISO`:
+
+  Durations are collapsed before they are applied:
+  - when shifting by 1 year and 2 months the date is actually shifted by 14 months
+  - weeks, days and smaller units are collapsed into seconds and microseconds
+
+  Durations are applied in order of the size of the unit: `month > day > second`.
+
+  ## Examples
+
+      iex> DateTime.shift(~U[2016-01-01 00:00:00Z], month: 2)
+      {:ok, ~U[2016-03-01 00:00:00Z]}
+      iex> DateTime.shift(~U[2016-01-01 00:00:00Z], year: 1, week: 4)
+      {:ok, ~U[2017-01-29 00:00:00Z]}
+      iex> DateTime.shift(~U[2016-01-01 00:00:00Z], minute: 25)
+      {:ok, ~U[2016-01-01 00:25:00Z]}
+      iex> DateTime.shift(~U[2016-01-01 00:00:00Z], minute: 5, microsecond: {500, 4})
+      {:ok, ~U[2016-01-01 00:05:00.0005Z]}
+  """
+  @doc since: "1.7.0"
+  @spec shift(
+          Calendar.datetime(),
+          Duration.t() | [Duration.unit()],
+          Calendar.time_zone_database()
+        ) ::
+          {:ok, t}
+          | {:ambiguous, first_datetime :: t, second_datetime :: t}
+          | {:gap, t, t}
+          | {:error,
+             :incompatible_calendars
+             | :time_zone_not_found
+             | :utc_only_time_zone_database}
+  def shift(datetime, duration, time_zone_database \\ Calendar.get_time_zone_database())
+
+  def shift(%{calendar: calendar} = datetime, %Duration{} = duration, time_zone_database) do
+    %{
+      year: year,
+      month: month,
+      day: day,
+      hour: hour,
+      minute: minute,
+      second: second,
+      microsecond: microsecond,
+      time_zone: time_zone
+    } = datetime
+
+    {year, month, day, hour, minute, second, microsecond} =
+      calendar.shift_naive_datetime(
+        year,
+        month,
+        day,
+        hour,
+        minute,
+        second,
+        microsecond,
+        duration
+      )
+
+    from_naive(
+      %NaiveDateTime{
+        calendar: calendar,
+        year: year,
+        month: month,
+        day: day,
+        hour: hour,
+        minute: minute,
+        second: second,
+        microsecond: microsecond
+      },
+      time_zone,
+      time_zone_database
+    )
+  end
```

Comment:
^ i am afraid this could ignore DST changes, which isn't necessarily a bad thing. Alternatively, DateTime.shift/3 could do the following:

1) shift_zone datetime to "Etc/UTC"
2) shift as naive datetime
3) from_naive/3 as "Etc/UTC"
4) shift_zone datetime to original time_zone

thoughts?

Reply:
I don't think it would ignore? I think the operation is semantically correct: you shift the wall clock and then you call `from_naive`, which will check if the time exists or not. In fact, maybe we should say precisely that in the docs. :)

@lau will be the person to confirm this.

Reply:
Yeah this could be reformulated at least: https://github.com/elixir-lang/elixir/pull/13385/files#diff-b83c142c2405080dc98ca11f1da9784fe7199a376c974e456a70b86c14bfd1a9R1684-R1686 :)